lib/rpc/handler/FallbackHandler.ts

```diff
@@ -0,0 +1,97 @@
+import Logger from 'bunyan'
+import { ChainId } from '@uniswap/sdk-core'
+import { getProviderId } from '../utils'
+import { ProviderHealthStateRepository } from '../ProviderHealthStateRepository'
+import { ProviderHealthStateDynamoDbRepository } from '../ProviderHealthStateDynamoDbRepository'
+import { ProviderHealthState } from '../ProviderHealthState'
+import { MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { metricScope, MetricsLogger } from 'aws-embedded-metrics'
+import { APIGatewayProxyResult } from 'aws-lambda'
+import { AWSMetricsLogger } from '../../handlers/router-entities/aws-metrics-logger'
+
+interface AlarmEvent {
+  alarmName: string
+  state: string
+  previousState: string
+  providerId: string
+  reason: string
+}
+
+export class FallbackHandler {
+  private healthStateRepository: ProviderHealthStateRepository
+  private log: Logger
+
+  constructor(log: Logger) {
+    const dbTableName = process.env.PROVIDER_HEALTH_STATE_DB_TABLE_NAME
+    if (!dbTableName) {
+      throw new Error('Missing DB_TABLE_NAME in env var')
+    }
+    this.healthStateRepository = new ProviderHealthStateDynamoDbRepository(dbTableName, log)
+    this.log = log
+  }
+
+  private buildHandler() {
+    return metricScope((metricsLogger: MetricsLogger) => async (event: object): Promise<APIGatewayProxyResult> => {
+      metricsLogger.setNamespace('Uniswap')
+      metricsLogger.setDimensions({ Service: 'RoutingAPI' })
+      const metric = new AWSMetricsLogger(metricsLogger)
+
+      const alarmEvent = this.readAlarmEvent(event)
+      this.log.debug({ alarmEvent }, 'Parsed alarmEvent')
+
+      if (
+        (alarmEvent.previousState === 'OK' || alarmEvent.previousState === 'INSUFFICIENT_DATA') &&
+        alarmEvent.state === 'ALARM'
+      ) {
+        metric.putMetric(`RPC_GATEWAY_FALLBACK_${alarmEvent.providerId}_INTO_UNHEALTHY`, 1, MetricLoggerUnit.Count)
+        this.log.error(
+          `${alarmEvent.providerId} becomes UNHEALTHY due to ${alarmEvent.previousState}=>ALARM in ${alarmEvent.alarmName}`
+        )
+        await this.healthStateRepository.write(alarmEvent.providerId, ProviderHealthState.UNHEALTHY)
```

Comment:
should this line come before the metric and log statement, in case it fails ?

Reply:
Nice suggestion! Fixed.